### PR TITLE
Virt notif watchdog

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This pull request adds support for notifications with nonsecure virtualization enabled.
This PR needs https://patchew.org/Xen/20240529072559.2486986-1-jens.wiklander@linaro.org/ for notifications to become enabled, but it's backward compatible so nothing should break.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
I believe the checkpatch issues are false positives.

File: core/arch/arm/kernel/thread_spmc.c [hunk 1]
```diff
@@ -1581,43 +1650,197 @@ static void handle_notification_get(struct thread_smc_args *args)
...
+static bool add_nvb_to_state(struct notif_info_get_state *state,
+			     uint16_t guest_id, struct notif_vm_bitmap *nvb)
+{
+	if (!nvb->pending)
+		return true;
+	/*
+	 * Add only the guest_id, meaning a global notification for this
+	 * guest.
+	 *
+	 * If notifications for a one or more specific vCPUs we'd add those
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread_spmc.c:1780]: "for **~~a~~** one or more"

File: core/arch/arm/kernel/virtualization.c [hunk 2]
```diff
@@ -409,20 +446,14 @@ TEE_Result virt_set_guest(uint16_t guest_id)
...
-		}
-	}
-	cpu_spin_unlock_xrestore(&prtn_list_lock, exceptions);
+	prtn = virt_get_guest(guest_id);
+	if (!prtn)
+		return TEE_ERROR_ITEM_NOT_FOUND;
 
-	return TEE_ERROR_ITEM_NOT_FOUND;
+	set_current_prtn(prtn);
+	core_mmu_set_prtn(prtn->mmu_prtn);
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/virtualization.c:556]: We have foreign interrupts masked at this point and we're only changing core-specific data. 
Yes, that's OK.

**Comment by jenswi-linaro** [comment]
Addressing comments.

**Comment by jenswi-linaro** [comment]
Rebased to fix CI issue.

**Comment by ldts** [comment]
@jenswi-linaro just curious, how are you testing? could you describe your setup? Would be great to have a doc mentioning Xen.
I am OoO for a few weeks and was toying around with Xen on Versal ACAP - maybe I can replicate on real hardware.

**Comment by jenswi-linaro** [comment]
We plan to have a working CI setup for this once the needed patches have been released in Xen version 4.19 and Linux v6.10.

You'll need:
- upstream Xen master containing commit b490f470f58d ("xen/arm: ffa: support notification")
- A Linux kernel v6.10 RC containing commit 0370fb127ce3 ("firmware: arm_ffa: Fix kernel warning about incorrect SRI/
NPI")

Xen must be configured with CONFIG_FFA=y and each guest have tee="ffa" in their config file.

OP-TEE should be configured with:
CFG_NS_VIRTUALIZATION=y CFG_CORE_SEL1_SPMC=y CFG_CORE_ASYNC_NOTIF=y
CFG_NOTIF_TEST_WD=y CFG_CORE_ASYNC_NOTIF_GIC_INTID ?= 8

TF-A need be configured for SPMC at S-EL1 with
SPD=spmd CTX_INCLUDE_EL2_REGS=0 SPMD_SPM_AT_SEL2=0
QEMU_TOS_FW_CONFIG_DTS=.../spmc_el1_manifest.dts SPMC_OPTEE=1

When booting you should notice lines like:
`D/TC:0 0   periodic_callback:136 seconds 1 millis 62 count 1`
Indicating that the periodic callback works

When DOM0 has booted you should see lines like:
```
D/TC:0 0   test_wd_callback:46 WD 1 call_count 1, timeout_count 0
D/TC:1 ? 0 wd_ndrv_yielding_cb:85 Clearing pending
```
Where `0 0  ` indicates test_wd_callback is called in an atomic context.
`1 ? 0` indicates that wd_ndrv_yielding_cb is called in a thread context of guest 1 (increased by 1 compared to what Xen is using).

You may want to test this on QEMU first (don't forget to update the optee_os, linux, and xen gits) to get the hang of it:
- `make XEN_BOOT=y SPMC_AT_EL=1 run`

**Comment by jenswi-linaro** [comment]
It would be nice to get this merged into the release.

**Comment by etienne-lms** [comment]
Looks consistent to me. I'm fine you squash the fixup commits.

**Comment by jenswi-linaro** [comment]
Squashed in the fixup commits.

File: core/kernel/notif.c [hunk 3]
```diff
@@ -57,31 +80,42 @@ void notif_unregister_driver(struct notif_driver *ndrv)
...
 	old_itr_status = cpu_spin_lock_xsave(&notif_lock);
 
-	if (notif_started) {
-		DMSG("Already started");
-		goto out;
+	switch (ev) {
+	case NOTIF_EVENT_STARTED:
+		if (ndata->notif_started) {
+			DMSG("Already started");
+			return;
```
Comment (etienne-lms) [inline core/kernel/notif.c:96]: `virt_put_guest()` to balance line 85?

**Review by etienne-lms — COMMENTED** [review]
Comment for commit "core: make generic notifications virtualization-aware".

File: core/kernel/notif.c (hunk 3 above)
Comment (jenswi-linaro) [reply core/kernel/notif.c]: Indeed, I'll add a goto out, we must unlock the spinlock too.

**Comment by jenswi-linaro** [comment]
Update

**Comment by etienne-lms** [comment]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for the series (with my comment addressed, IIUC).
All checkpatch warning & error reports look false positive to me.

**Comment by jenswi-linaro** [comment]
Thanks, tag applied.

**Comment by jenswi-linaro** [comment]
There's that pkcs11_1007 FAIL again. We should look into that error before the release.

**Comment by jforissier** [comment]
> There's that pkcs11_1007 FAIL again. We should look into that error before the release.

I could not reproduce it in my test setup. We may have to modify the CI script a bit to show the detailed logs on error.

**Pull request merged**